odl/solvers/nonsmooth/proximal_operators.py

```diff
@@ -1441,6 +1443,231 @@ def _call(self, x, out):
+    proj_simplex : projection onto simplex
+    """
+
+    if out is None:
+        out = x.space.element()
+
+    u = x.ufuncs.absolute()
+    v = x.ufuncs.sign()
+    proj_simplex(u, radius, out)
+    out *= v
+
+    return out
+
+
+def proj_simplex(x, radius=1, out=None):
+    """Projection onto simplex.
+
+    Projection onto::
+
+        ``{ x \in X | x_i \geq 0, \sum_i x_i = r}``
+
+    with `r` being the radius.
+
+    Parameters
+    ----------
+    space : `LinearSpace`
+        Space / domain `X`.
+    radius : positive float, optional
+        Radius `r` of simplex.
+
+    Returns
+    -------
+    prox_factory : callable
+        Factory for the proximal operator to be initialized.
+
+    Notes
+    -----
+    The projection onto a simplex is not of closed-form but can be solved by a
+    non-iterative algorithm, see [D+2008] for details.
+
+    References
+    ----------
+    [D+2008] Duchi, J., Shalev-Shwartz, S., Singer, Y., and Chandra, T.
+    *Efficient Projections onto the L1 -ball for Learning in High dimensions*.
+    ICML 2008, pp. 272–279. http://doi.org/10.1145/1390156.1390191
+
+    See Also
+    --------
+    proj_l1 : projection onto l1-ball
+    """
+
+    if out is None:
+        out = x.space.element()
+
+    # sort values in descending order
+    x_vec = x.asarray().flatten()
+    x_sor = np.sort(x_vec)[::-1]
+
+    # TODO: explain a little what gets computed here
+    j = np.arange(1, x.size + 1)
+    x_avrg = (1 / j) * (np.cumsum(x_sor) - radius)
+    crit = x_sor - x_avrg
+    i = np.argwhere(crit >= 0).flatten().max()
+    out[:] = np.maximum(x - x_avrg[i], 0)
+
+    return out
+
+
+def proximal_simplex(space, radius=1):
+    """Proximal operator factory of the characteristic function of simplex.
+
+    Function for the proximal operator of the functional ``F`` where ``F``
+    is
+
+    .. math::
+        F(x) &=  \chi_{S^{n-1}_r}(x)
+             &= \\begin{cases} 0 & \\text{if } x \in S^{n-1}_r \\\\
+                               \infty & \\text{else},
+        \\end{cases}
+
+    and `S^{n-1}_r = { x \in \R^n | x_i \geq 0, \sum_{i=1}^n x_i = r }` is the
+    simplex of radius `r`.
+
+    Parameters
+    ----------
+    space : `LinearSpace`
+        Domain of F(x).
+    radius : positive float, optional
+        Radius `r` of simplex.
+
+    Returns
+    -------
+    prox_factory : callable
+        Factory for the proximal operator to be initialized.
+
+    Notes
+    -----
+    The proximal operator is not of closed-form but can be computed by a
+    non-iterative algorithm, see [D+2008] for details.
+
+    References
+    ----------
+    [D+2008] Duchi, J., Shalev-Shwartz, S., Singer, Y., and Chandra, T.
+    *Efficient Projections onto the L1 -ball for Learning in High dimensions*.
+    ICML 2008, pp. 272–279. http://doi.org/10.1145/1390156.1390191
+
+    See Also
+    --------
+    proj_simplex : projection onto simplex
+    """
+
+    class ProxSimplex(Operator):
```

Comment:
Indeed, I'd try doing that.

Overall the split on ODL is mostly for historical reasons, we could probably move lots of stuff around behind the scenes.

Reply:
Yes, for the relatively simple cases we should in the future just inline the class in the `functional.proximal` property. No need for an extra indirection.
We should even consider removing these factories altogether and put the parameters in the `__init__` of the prox operators. The functional property can then forward the arguments in the correct way.

The current situation is only because the initial solvers interface used the prox operators directly, and therefore they had to have a uniform interface.